RFC 2008: Future-proofing enums/structs with #[non_exhaustive] attribute

davidtwco opened the pull request:
This work-in-progress pull request contains my changes to implement [RFC 2008](https://github.com/rust-lang/rfcs/pull/2008). The related tracking issue is #44109.

As of writing, enum-related functionality is not included and there are some issues related to tuple/unit structs. Enum related tests are currently ignored.

WIP PR requested by @nikomatsakis [in Gitter](https://gitter.im/rust-impl-period/WG-compiler-middle?at=59e90e6297cedeb0482ade3e).

rust-highfive commented:
Thanks for the pull request, and welcome! The Rust team is excited to review your changes, and you should hear from @steveklabnik (or someone else) soon.

If any changes to this PR are deemed necessary, please add them as extra commits. This ensures that the reviewer can see what has changed since they last reviewed the code. Due to the way GitHub handles out-of-date commits, this should also make it reasonably obvious what issues have or haven't been addressed. Large or tricky changes may require several passes of review and changes.

Please see [the contribution instructions](https://github.com/rust-lang/rust/blob/master/CONTRIBUTING.md) for more information.

nikomatsakis commented:
r? @nikomatsakis

nikomatsakis commented on src/librustc/ty/mod.rs:
```diff
@@ -1371,6 +1371,7 @@ pub struct AdtDef {
     pub variants: Vec<VariantDef>,
     flags: AdtFlags,
     pub repr: ReprOptions,
+    pub non_exhaustive: bool,
```
👍

nikomatsakis replied:
Can you add a comment briefly explaining what this is, perhaps linking to the RFC? I know none of the other fields have comments, but that's no reason not to do better... =)

davidtwco replied:
I definitely will, I plan on doing another pass to make sure everything is well commented, just wanted to get that pushed up before I called it a night. Will do this ASAP.

davidtwco replied:
Just pushed up a comment here.

oli-obk commented on src/test/compile-fail/rfc-2008-non-exhaustive/enum_field_attempt_create.rs:
```diff
@@ -0,0 +1,30 @@
+use enums::FieldEnum;
+
+// This errors as we cannot create a enum variant marked non_exhaustive
+// from an external crate as if the create added new fields it would be
```
typo: `create` -> `crate`

davidtwco replied:
Fixed.

oli-obk commented on src/test/compile-fail/rfc-2008-non-exhaustive/enum_missing_wildcard.rs:
```diff
@@ -0,0 +1,34 @@
+
+fn match_enum(ne: &NormalEnum) -> &str {
+    match *ne {
+        //~^ ERROR `..` required with variant marked as non-exhaustive
```
The error message is wrong, the `enum` is marked as non-exhaustive, not the variant

davidtwco replied:
Fixed this part of the test.

oli-obk commented on src/librustc_typeck/check/_match.rs:
```diff
@@ -867,6 +868,15 @@ impl<'a, 'gcx, 'tcx> FnCtxt<'a, 'gcx, 'tcx> {
+        if tcx.sess.features.borrow().non_exhaustive {
+            // Require `..` if structure or enum has non_exhaustive attribute.
+            if adt.non_exhaustive && !adt.did.is_local() && !etc {
+                tcx.sess.span_err(
```
I think this error should have its own error number.

davidtwco replied:
I've added error numbers in a WIP commit that I couldn't get compiling locally - messages regarding the diagnostic code not being registered despite my not being able to spot a difference in how they are being added. Would appreciate any thoughts here.

oli-obk commented on src/librustc_typeck/check/mod.rs:
```diff
@@ -3408,6 +3408,18 @@ impl<'a, 'gcx, 'tcx> FnCtxt<'a, 'gcx, 'tcx> {
+                let descr = self.tcx.adt_def(variant.did).variant_descr();
+
+                // Prohibit struct expressions when non exhaustive flag is set.
+                self.tcx.sess.span_err(expr.span, &format!("cannot create non-exhaustive {}",
```
Same here, but a different new error code

davidtwco replied:
See other comment w/r/t error codes.

oli-obk commented:
tidy failed:

```
[00:03:26] tidy error: /checkout/src/test/compile-fail/rfc-2008-non-exhaustive/functional_record.rs:24: line longer than 100 chars
[00:03:26] tidy error: /checkout/src/librustc_privacy/lib.rs:632: line longer than 100 chars
[00:03:26] tidy error: /checkout/src/librustc_privacy/lib.rs:635: line longer than 100 chars
```

davidtwco commented:
@oli-obk Thanks for the comments, I'll get on these ASAP.

oli-obk commented on src/librustc/diagnostics.rs:
````diff
@@ -1939,6 +1939,65 @@ fn foo<'a>(x: &'a i32, y: &i32) -> &'a i32 {
 ```
 "##,
 
+E0638: r##"
````
You need to add this to the diagnostics file of librustc_typeck

davidtwco replied:
Thanks, I've pushed up a fix for this. I wasn't aware that there was a diagnostics file in the other crates, my bad.

petrochenkov commented on src/librustc/ty/mod.rs:
```diff
@@ -1371,6 +1371,12 @@ pub struct AdtDef {
+    /// fields/variants of this data type.
+    ///
+    /// See RFC 2008 (https://github.com/rust-lang/rfcs/pull/2008).
+    pub non_exhaustive: bool,
```
This is better merged into the `flags: AdtFlags` field.

davidtwco replied:
Fixed.

petrochenkov commented on src/librustc_metadata/encoder.rs:
```diff
@@ -648,6 +649,14 @@ impl<'a, 'b: 'a, 'tcx: 'b> IsolatedEncoder<'a, 'b, 'tcx> {
             }
         }
 
+        if tcx.sess.features.borrow().non_exhaustive {
```
This condition seems unnecessary.

davidtwco replied:
Fixed.

petrochenkov commented on src/librustc_privacy/lib.rs:
```diff
@@ -625,6 +625,18 @@ impl<'a, 'tcx> TypePrivacyVisitor<'a, 'tcx> {
                             }
                         }
+
+                        if self.tcx.sess.features.borrow().non_exhaustive {
```
Unnecessary condition again.

davidtwco replied:
Fixed.

petrochenkov commented on src/librustc_resolve/build_reduced_graph.rs:
```diff
@@ -337,8 +337,20 @@ impl<'a> Resolver<'a> {
 
             // These items live in both the type and value namespaces.
             ItemKind::Struct(ref struct_def, _) => {
+                if self.session.features.borrow().non_exhaustive {
```
Unnecessary condition again, you don't have to check this all the time, one feature check in `feature_gate.rs` is enough.

davidtwco replied:
Fixed. Wasn't aware that I didn't need it everywhere - I misinterpreted the instructions on the [Rust forge](https://forge.rust-lang.org/feature-guide.html).

petrochenkov commented on src/librustc_resolve/build_reduced_graph.rs:
```diff
@@ -337,8 +337,20 @@ impl<'a> Resolver<'a> {
+                    // If the structure is marked as non_exhaustive then lower the visibility
+                    // to within the crate.
+                    if has_non_exhaustive && vis == ty::Visibility::Public {
+                        vis = ty::Visibility::Restricted(DefId::local(CRATE_DEF_INDEX));
```
Hmm, this doesn't look correct - constructor's visibility below (`ctor_vis`) should be lowered to `pub(crate)`, not visibility of the struct itself.

davidtwco replied:
Adjusted this.

petrochenkov commented on src/librustc_typeck/check/_match.rs:
```diff
@@ -867,6 +868,15 @@ impl<'a, 'gcx, 'tcx> FnCtxt<'a, 'gcx, 'tcx> {
             self.check_pat_walk(&field.pat, field_ty, def_bm, true);
         }
 
+        if tcx.sess.features.borrow().non_exhaustive {
```
Unnecessary feature check.

davidtwco replied:
Fixed.

davidtwco replied:
After doing this, I noticed a regression in creating the structs - wasn't able to work out what was causing it.

petrochenkov commented on src/librustc_typeck/check/mod.rs:
```diff
@@ -3408,6 +3408,19 @@ impl<'a, 'gcx, 'tcx> FnCtxt<'a, 'gcx, 'tcx> {
             return self.tcx.types.err;
         };
 
+        if self.tcx.sess.features.borrow().non_exhaustive {
```
Unnecessary feature check.

davidtwco replied:
Fixed.

petrochenkov commented on src/librustc_typeck/check/_match.rs:
```diff
@@ -867,6 +868,15 @@ impl<'a, 'gcx, 'tcx> FnCtxt<'a, 'gcx, 'tcx> {
         }
 
+        if tcx.sess.features.borrow().non_exhaustive {
+            // Require `..` if structure or enum has non_exhaustive attribute.
```
Require `..` if structure or *variant* has the `non_exhaustive` attribute.
It looks like this patch doesn't implement `non_exhaustive` for variants yet, but `non_exhaustive` on `enum` should not restrict use of its variants.

davidtwco replied:
Adjusted this comment.

petrochenkov commented on src/librustc_typeck/check/mod.rs:
```diff
@@ -3408,6 +3408,19 @@ impl<'a, 'gcx, 'tcx> FnCtxt<'a, 'gcx, 'tcx> {
+        if self.tcx.sess.features.borrow().non_exhaustive {
+            let is_local = variant.did.is_local();
+
+            if !is_local {
```
Looks like the check for `non_exhaustive` is forgotten here.

petrochenkov replied:
Also see https://github.com/rust-lang/rust/pull/45394/files#r146130454, `non_exhaustive` on `enum` should not prohibit struct expressions using that enum's variants.

davidtwco replied:
Added check.

petrochenkov commented on src/test/compile-fail/rfc-2008-non-exhaustive/auxiliary/enums.rs:
```diff
@@ -0,0 +1,24 @@
+
+pub enum FieldEnum {
+    WithoutFields,
+    #[non_exhaustive] WithAnonStruct { field: u32 },
```
Since `non_exhaustive` on variants is not implemented, could you make it an error? (Currently it's silently ignored.)

davidtwco replied:
I've not gotten around to doing this yet.

petrochenkov commented on src/test/compile-fail/rfc-2008-non-exhaustive/auxiliary/enums.rs:
```diff
@@ -0,0 +1,24 @@
+pub enum FieldEnum {
+    WithoutFields,
+    #[non_exhaustive] WithAnonStruct { field: u32 },
+    #[non_exhaustive] WithTuple(u32)
```
I'd like to see all of these configurations tested:
```
#[non_exhaustive]
enum NonExhaustiveEnum {
    Unit,
    Tuple(u8),
    Struct { field: u8 },
}

enum NonExhaustiveVariants {
    #[non_exhaustive] Unit,
    #[non_exhaustive] Tuple(u8),
    #[non_exhaustive] Struct { field: u8 },
}
```

davidtwco replied:
Now testing with these enums.

petrochenkov commented on src/test/compile-fail/rfc-2008-non-exhaustive/auxiliary/structs.rs:
```diff
@@ -0,0 +1,37 @@
+// option. This file may not be copied, modified, or distributed
+// except according to those terms.
+
+#![feature(non_exhaustive)]
```
Unnecessary feature, `feature(non_exhaustive)` is needed only in crates defining `non_exhaustive` things.

petrochenkov replied:
Whoops, wrong place for the comment.

petrochenkov commented on src/test/compile-fail/rfc-2008-non-exhaustive/enum_field_attempt_create.rs:
```diff
@@ -0,0 +1,30 @@
+// option. This file may not be copied, modified, or distributed
+// except according to those terms.
+
+// FIXME issue #44109, PR #45394
```
Nit: This can probably be omitted (also FIXME seems out-of-place).

petrochenkov replied:
Same with other tests.

davidtwco replied:
Done.

petrochenkov commented on src/test/compile-fail/rfc-2008-non-exhaustive/enum_field_attempt_create.rs:
```diff
@@ -0,0 +1,30 @@
+// except according to those terms.
+
+// FIXME issue #44109, PR #45394
+// gate-test-non_exhaustive
```
This marker is used for feature gate tests, but this test is not a feature gate test.

Also, could you actually add a feature gate test? (Making sure a feature error is reported in `#[non_exhaustive]` is used without `feature(non_exhaustive)`).

petrochenkov replied:
Same with other tests.

davidtwco replied:
I've added a proper feature gate test.

petrochenkov commented on src/test/compile-fail/rfc-2008-non-exhaustive/enum_field_attempt_create.rs:
```diff
@@ -0,0 +1,30 @@
+
+// FIXME issue #44109, PR #45394
+// gate-test-non_exhaustive
+#![feature(non_exhaustive)]
```
Unnecessary feature, `feature(non_exhaustive)` is required only in crated defining something `non_exhaustive`, not using it.

petrochenkov replied:
Same with other tests.

davidtwco replied:
Removed the unnecessary feature usages.

petrochenkov commented on src/librustc_typeck/check/mod.rs:
```diff
@@ -3408,6 +3408,19 @@ impl<'a, 'gcx, 'tcx> FnCtxt<'a, 'gcx, 'tcx> {
+
+                // Prohibit struct expressions when non exhaustive flag is set.
+                span_err!(self.tcx.sess, expr.span, E0639,
+                          "cannot create non-exhaustive {}",
```
Nit: `"cannot create non-exhaustive {} using struct expression"`?
I mean you can obtain non-exhaustive structs using other means, e.g. function calls.

davidtwco replied:
Fixed.

petrochenkov commented on src/test/compile-fail/rfc-2008-non-exhaustive/unit_struct_match_exhaustively.rs:
```diff
@@ -0,0 +1,27 @@
+
+fn match_struct(us: &UnitStruct) {
+    let &UnitStruct { } = us;
+    //~^ ERROR `..` required with struct marked as non-exhaustive
```
You can merge several test files testing similar things into one file with several `//~ ERROR` markers, it's usually more convenient that way (both for writing better more exhaustive tests, and for reviewing them).

davidtwco replied:
Changed this. It's a lot easier to keep track of now.

davidtwco commented:
@petrochenkov Thanks a bunch for the comments. Just fixed a bunch of them and I think there's one that I'm still working to resolve.

arielb1 commented on src/librustc_const_eval/_match.rs:
```diff
@@ -628,10 +642,15 @@ pub fn is_useful<'p, 'a: 'p, 'tcx: 'a>(cx: &mut MatchCheckCtxt<'a, 'tcx>,
         debug!("missing_ctors={:?} is_privately_empty={:?}", missing_ctors,
                is_privately_empty);
+
+        if cx.is_non_exhaustive(pcx.ty) && !cx.is_local(pcx.ty) {
```
It's confusing to mutate this variable, especially *after* the debug println

Maybe have
```Rust
let is_privately_empty =
    all_ctors.is_empty() && !cx.is_uninhabited(pcx.ty);
let is_declared_nonexhaustive =
    cx.is_non_exhaustive(pcx.ty) && !cx.is_local(pcx.ty);

// For privately empty and nonexhaustive enums, we work as
// if there were an "extra" `_` constructor for the type, so we
// can never match over all constructors.
let is_nonexhaustive =
    is_privately_empty || is_declared_nonexhaustive;
if missing_ctors.is_empty() && !is_nonexhaustive
```

davidtwco replied:
Fixed this, thanks.

arielb1 commented:
There's invalid syntax in your run-pass test:
```
[01:07:10] error: expected expression, found `}`
[01:07:10]   --> /checkout/src/test/run-pass/rfc-2008-non-exhaustive/enums.rs:17:67
[01:07:10]    |
[01:07:10] 17 |     let variant_tuple = NonExhaustiveVariants::Tuple { 0: 340, .. };
[01:07:10]    |                                                                   ^
[01:07:10] 
[01:07:10] error: expected expression, found `}`
[01:07:10]   --> /checkout/src/test/run-pass/rfc-2008-non-exhaustive/enums.rs:18:73
[01:07:10]    |
[01:07:10] 18 |     let variant_struct = NonExhaustiveVariants::Struct { field: 340, .. };
[01:07:10]    |                                                                         ^

[01:07:10]   --> /checkout/src/test/run-pass/rfc-2008-non-exhaustive/structs.rs:17:69
[01:07:10]    |
[01:07:10] 17 |     let ns = NormalStruct { first_field: 640, second_field: 480, .. };
[01:07:10]    |                                                                     ^
[01:07:10] 
[01:07:10] error: expected expression, found `}`
[01:07:10]   --> /checkout/src/test/run-pass/rfc-2008-non-exhaustive/structs.rs:21:47
[01:07:10]    |
[01:07:10] 21 |     let ts = TupleStruct { 0: 340, 1: 480, .. };
[01:07:10]    |                                               ^
[01:07:10] 
[01:07:10] error: expected expression, found `}`
[01:07:10]   --> /checkout/src/test/run-pass/rfc-2008-non-exhaustive/structs.rs:25:30
[01:07:10]    |
[01:07:10] 25 |     let us = UnitStruct { .. };
```

You can't do FRU on enums, so there's nothing to test, but for structs, you should specify a base:
```Rust
let base = UnitStruct;
let us = UnitStruct { ..base };
```

etc.

davidtwco commented:
@arielb1 I've modified the `run-pass` tests so that they don't have syntax errors. There is a unexpected panic happening now that I'll need to figure out what is causing. I'm still unsure as to why none of the `compile-fail` test errors are happening as I was definitely seeing those errors when adjusting the `run-pass` tests.

davidtwco commented:
@arielb1 @petrochenkov Thanks for all previous comments.

I've been struggling to make further progress on this.

I've narrowed down the compiler panic in the `run-pass` tests to the change made to the `is_privately_empty` code in [src/librustc_const_eval/_match.rs](src/librustc_const_eval/_match.rs#L645-L654) but I've not made much progress in fixing it.

Most of the `compile-fail` tests seem to be related to constructor visibilities not being made lower as far as I can tell. I've added a check to `fn def_id_visibility` in `librustc_privacy` as mentioned in the initial mentoring instructions but that only seems to reduce the variant's visibility (not just the constructor). I also added code to `build_reduced_graph_for_variant` in `librustc_resolve` to lower the constructor visibility but it doesn't seem to be making a difference. I've just pushed up a commit with that.

I also experimented locally in implementing a `is_non_exhaustive` field on the `VariantDef` as was mentioned in Gitter at some point but I didn't find it particularly useful in the areas where I've been messing with constructor visibilities.

Would appreciate any input as to where I'm going wrong, and apologies for how long this PR is taking.

As an aside, I'd appreciate some advice on debugging the compiler - I've tried adding `debug!` calls and then setting `RUST_LOG=debug` but that doesn't seem to apply to the tests when run from `x.py` (also looked into the `config.toml` file for this). I've also used `RUST_BACKTRACE` but I found it tough to determine exactly where the calls were compared to traces that I'm familiar with in other tools/languages.

petrochenkov commented:
@davidtwco 
>As an aside, I'd appreciate some advice on debugging the compiler

`RUST_LOG` output is enabled by `debug-assertions = true` in `config.toml`.
You can run the produced compiler without `x.py` like this `build/x86_64-pc-windows-gnu/stage1/bin/rustc path/to/my/test.rs`, then output from `RUST_LOG` won't be suppressed.

>I also experimented locally in implementing a is_non_exhaustive field on the VariantDef

As I mentioned in https://github.com/rust-lang/rust/issues/44109#issuecomment-329962555, `non_exhaustive` on variants may have some infrastructural issues due to separating visibilities for variants and their constructors, so they can be omitted from the initial implementation.

>Most of the compile-fail tests seem to be related to constructor visibilities not being made lower as far as I can tell.

I don't see any tests checking availability of constructors in this PR. Could you add them with `ERROR` annotations corresponding to errors produced by compiler on this branch right now? (So all the compile-fail tests pass as a result).
The implementation looks correct.

arielb1 commented:
> I've also used `RUST_BACKTRACE` but I found it tough to determine exactly where the calls were compared to traces that I'm familiar with in other tools/languages.

Try setting `debuginfo-lines = true` in your config.toml, and rebase on top of the current nightly (it will be much faster), then you'll get line numbers on RUST_BACKTRACE.

davidtwco commented:
Thanks @arielb1and @petrochenkov. I've fixed the compiler panic that was causing issues with the `run-pass` tests now and rebased onto master.

@petrochenkov I'm pretty sure I am misunderstanding or unaware of something regarding: 

> I don't see any tests checking availability of constructors in this PR. Could you add them with ERROR annotations corresponding to errors produced by compiler on this branch right now? (So all the compile-fail tests pass as a result).

Is this not [what this line](https://github.com/davidtwco/rust/blob/aa32fdbba43e8926acee2220ffa634d589435c1e/src/test/compile-fail/rfc-2008-non-exhaustive/structs.rs#L25) and lines 31 and 37 are testing?

petrochenkov commented:
@davidtwco 
>Is this not what this line and lines 31 and 37 are testing?

No, those are "struct expressions" or "struct literals".
Constructors are
```rust
let ts = TupleStruct(640, 480); // Function-like constructor of a tuple struct
let us = UnitStruct; // Constant-like constructor of a unit struct

// Plus same for patterns
match ... {
    TupleStruct(..) => {}
}
match ... {
    UnitStruct => {}
}
```
, the privacy tweaks affect only them actually, not braced forms.

davidtwco commented:
@petrochenkov Thanks for the clarification. I've found when changing it to:

```rust
let ts = TupleStruct(640, 480); // Function-like constructor of a tuple struct
let us = UnitStruct; // Constant-like constructor of a unit struct
```

that I get:

```
unexpected errors (from JSON output): [
    Error {
        line_num: 30,
        kind: Some(
            Error
        ),
        msg: "30:14: 30:25: expected function, found struct `TupleStruct` [E0423]"
    },
    Error {
        line_num: 36,
        kind: Some(
            Error
        ),
        msg: "36:14: 36:24: expected value, found struct `UnitStruct` [E0423]"
    }
]
```

which is why I was using the struct expression form in the tests. I assume that this must be caused by a change I've made rather than a misunderstanding I might have of the syntax?

petrochenkov commented:
@davidtwco 
Those errors are exactly the expected effect from privatizing constructors - now other crates don't know about their existence (unless full paths are used, then you get a privacy error, but that's  a rarer situation).

davidtwco commented:
@petrochenkov Ah, fantastic! It seems I've spent far too long trying to figure out what I was doing wrong with those errors while it was actually working. I've pushed up a change to the `ERROR` comments to reflect this.

nikomatsakis commented:
@davidtwco great! So what is current status of this PR, are there remaining work items?

davidtwco commented:
@nikomatsakis I think there is only one pending issue. There's no error [on this line](https://github.com/davidtwco/rust/blob/eb6395fddf5dfb7a4b9526b0242b036cd20da439/src/test/compile-fail/rfc-2008-non-exhaustive/structs.rs#L25) which I believe is wrong.

[These tests](https://github.com/davidtwco/rust/blob/eb6395fddf5dfb7a4b9526b0242b036cd20da439/src/test/compile-fail/rfc-2008-non-exhaustive/enums.rs) also fail, however I believe that is expected as we're not handling variant constructors.

That's it unless anyone else spots anything I've missed.

petrochenkov commented:
@davidtwco 
>There's no error on this line which I believe is wrong.

The check in `src/librustc_typeck/check/mod.rs` needs to moved a few lines higher, now it is run only when there's `..base` in the struct expression.

arielb1 commented on src/librustc_const_eval/_match.rs:
```diff
@@ -630,12 +644,21 @@ pub fn is_useful<'p, 'a: 'p, 'tcx: 'a>(cx: &mut MatchCheckCtxt<'a, 'tcx>,
             all_ctors.into_iter().map(|c| {
                 is_useful_specialized(cx, matrix, v, c.clone(), pcx.ty, witness)
             }).find(|result| result.is_useful()).unwrap_or(NotUseful)
+        } else if is_non_exhaustive {
```
This looks like it would prevent matching non-exhaustive patterns by *wildcard* patterns, e.g.
```Rust
match non_exhaustive {
    _ => {}
}
```

davidtwco replied:
I've pushed a fix for this. However, I am not 100% sure that it solves this issue, I'd appreciate any further comments.

davidtwco replied:
Added a test for my fix and it seems fine.

davidtwco commented:
@petrochenkov Thanks! Just pushed up this fix.

arielb1 commented on src/librustc_const_eval/_match.rs:
```diff
@@ -647,7 +668,10 @@ pub fn is_useful<'p, 'a: 'p, 'tcx: 'a>(cx: &mut MatchCheckCtxt<'a, 'tcx>,
                 UsefulWithWitness(pats) => {
                     let cx = &*cx;
-                    let new_witnesses = if used_ctors.is_empty() {
+                    // If we have a non exhaustive enum, then it is possible for execution to
```
better comment:
```Rust
// In this case, there's at least one "free"
// constructor that is only matched against by
// wildcard patterns.
//
// There are 2 ways we can report a witness here.
// Commonly, we can report all the "free"
// constructors as witnesses, e.g. if we have:
//
// ```
//     enum Direction { N, S, E, W }
//     let Direction::N = ...;
// ```
//
// we can report 3 witnesses: `S`, `E`, and `W`.
//
// However, there are 2 cases where we don't want
// to do this and instead report a single `_` witness:
//
// 1) If the user is matching against a non-exhaustive
// enum, there is no point in enumerating all possible
// variants, because the user can't actually match
// against them himself, e.g. in an example like:
// ```
//     let err: io::ErrorKind = ...;
//     match err {
//         io::ErrorKind::NotFound => {},
//     }
// ```
// we don't want to show every possible IO error,
// but instead have `_` as the witness (this is
// actually *required* if the user specified *all*
// IO errors, but is probably what we want in every
// case).
//
// 2) If the user didn't actually specify a constructor
// in this arm, e.g. in
// ```
//     let x: (Direction, Direction, bool) = ...;
//     let (_, _, false) = x;
// ```
// we don't want to show all 16 possible witnesses
// `(<direction-1>, <direction-2>, true)` - we are
// satisfied with `(_, _, true)`. In this case,
// `used_ctors` is empty.
```

davidtwco replied:
Thanks, that is better, added it with recent push.

arielb1 commented:
I'm cool with the `check_match` changes. I suppose this is only waiting for review by @petrochenkov ?

davidtwco commented:
Not sure why there are still errors with the doctests.

petrochenkov commented on src/librustc_const_eval/_match.rs:
```diff
@@ -208,6 +208,20 @@ impl<'a, 'tcx> MatchCheckCtxt<'a, 'tcx> {
         }
     }
 
+    fn is_non_exhaustive(&self, ty: Ty<'tcx>) -> bool {
```
Nit: slightly misleading name.
`is_non_exhaustive` -> `is_non_exhaustive_enum`

davidtwco replied:
Resolved.

petrochenkov commented on src/librustc_typeck/check/_match.rs:
```diff
@@ -867,6 +868,13 @@ impl<'a, 'gcx, 'tcx> FnCtxt<'a, 'gcx, 'tcx> {
         }
 
+        // Require `..` if structure or variant has non_exhaustive attribute.
+        if adt.is_non_exhaustive() && !adt.did.is_local() && !etc {
```
We are checking variant patterns here, but `adt.is_non_exhaustive()` is returns non-exhaustiveness of the variant's enum, not of the variant itself.
I.e. this will incorrectly generate an error.
```rust
// Crate 1
#[non_exhaustive]
enum E { V { a: u8 } }

// Crate 2
match ... {
    E::V { a } => {}
}
```
(Could you also add this case to tests?)

davidtwco replied:
Resolved.

petrochenkov commented on src/librustc/ty/mod.rs:
```diff
@@ -1370,7 +1376,7 @@ pub struct AdtDef {
     pub variants: Vec<VariantDef>,
     flags: AdtFlags,
-    pub repr: ReprOptions,
+    pub repr: ReprOptions
```
Tiny nit: comma has disappeared

davidtwco replied:
Resolved.

petrochenkov commented on src/librustc_resolve/build_reduced_graph.rs:
```diff
@@ -414,7 +425,18 @@ impl<'a> Resolver<'a> {
+            ctor_vis = ty::Visibility::Restricted(DefId::local(CRATE_DEF_INDEX));
+        }
+
+        self.define(parent, ident, ValueNS, (ctor_def, ctor_vis, variant.span, expansion));
```
Since we don't support variants yet, this change can be removed.

davidtwco replied:
Resolved.

petrochenkov commented:
@davidtwco 
Variants are not supported yet, but the attribute is still silently accepted on them.
Could you temporarily prohibit `#[non_exhaustive]` on variants in `ast_validation.rs`?

petrochenkov commented:
Almost ready, r=me after addressing the last round of comments and squashing commits.

davidtwco commented:
@petrochenkov That's all your comments resolved.

@bors r=petrochenkov

bors commented:
@davidtwco: :key: Insufficient privileges: Not in reviewers

davidtwco commented:
Oops, looks like I've accidently mangled some other commits while rebasing.

petrochenkov commented:
@bors r+

bors commented:
:clipboard: Looks like this PR is still in progress, ignoring approval

petrochenkov commented:
@bors r+

bors commented:
:pushpin: Commit d3babe5 has been approved by `petrochenkov`

<!-- @bors r=petrochenkov d3babe5eaa073b93c25529831ac8afb37f47c412 -->

bors commented:
:hourglass: Testing commit d3babe5eaa073b93c25529831ac8afb37f47c412 with merge 5fe8cd7aa66c92d1936d8286e6205e24e566aa7c...

bors commented:
:broken_heart: Test failed - [status-travis](https://travis-ci.org/rust-lang/rust/builds/297080928?utm_source=github_status&utm_medium=notification)

kennytm commented:
`pretty` test of `run-pass/rfc-2008-non-exhaustive/enums.rs` failed.

```
[01:16:34] ---- [pretty] run-pass/rfc-2008-non-exhaustive/enums.rs stdout ----
[01:16:34] 	
[01:16:34] error: pretty-printed source does not match expected source
```

Diff (a single empty line 😕):

```diff
--- expected
+++ actual
@@ -20,6 +20,7 @@
 [01:16:34]         NonExhaustiveEnum::Unit => 1,
 [01:16:34]         NonExhaustiveEnum::Tuple(_) => 2,
 [01:16:34] 
+[01:16:34] 
 [01:16:34]         // This particular arm tests that a enum marked as non-exhaustive
 [01:16:34]         // will not error if its variants are matched exhaustively.
 [01:16:34]         NonExhaustiveEnum::Struct { field } => field,
```

petrochenkov commented:
This is a pretty-printer bug (https://github.com/rust-lang/rust/issues/37199).
`// ignore-pretty issue #37199` in the test will fix the issue.

petrochenkov commented:
@bors r+

bors commented:
:pushpin: Commit 86c62d0 has been approved by `petrochenkov`

<!-- @bors r=petrochenkov 86c62d02eebb037faf7c0752a9c472181e5608cb -->

bors commented:
:hourglass: Testing commit 86c62d02eebb037faf7c0752a9c472181e5608cb with merge d762b1d6c67db12e117186d94d70e46cddb22965...

bors commented:
:sunny: Test successful - [status-appveyor](https://ci.appveyor.com/project/rust-lang/rust/build/1.0.5231), [status-travis](https://travis-ci.org/rust-lang/rust/builds/297288270?utm_source=github_status&utm_medium=notification)
Approved by: petrochenkov
Pushing d762b1d6c67db12e117186d94d70e46cddb22965 to master...

nikomatsakis commented:
Just saw that this landed, nice work @davidtwco !
